crossfire/__init__.py

```diff
@@ -52,15 +58,19 @@ def occurrences(
 
 
 def flatten(data, nested_columns=None):
+    origin_type = type(data)
     nested_columns = set(nested_columns or NESTED_COLUMNS)
     if not nested_columns.issubset(NESTED_COLUMNS):
         raise NestedColumnError(nested_columns)
-    if not data:
-        return data
-    if isinstance(data, list):
-        keys = set(data[0].keys()) & nested_columns
-        for item in data:
-            for key in keys:
-                item.update({f"{key}_{k}": v for k, v in item.get(key).items()})
-                item.pop(key)
+    if isinstance(data, DataFrame):
+        data = data.to_dict(orient="records")
```

Comment:
So, considering your [provocation](https://github.com/FelipeSBarros/crossfire/issues/96#issuecomment-1960282545) I thought it would be easier to make the test pass by converting `DataFrame` to `dcit`, to run the flatten process....

Reply:
I think it is too expensive to convert a DataFrame to dict to just convert it back to DataFrame later… my provocation showed you a way to deal with different logic depending on the type of the input without using `type(…)` or `isinstance(…)`. I am not sure what you meant here…

Reply:
> I think it is too expensive to convert a DataFrame to dict to just convert it back to DataFrame later… 

I could imagine that... but my goal was to make the test pass... :upside_down_face: 

It is evident that I dind't get your provocation, right? I understand that, you can build a function `add` accepting both, string and integers, by casting (not sure if this is the correct concept/vocalulary) whatever comes to `int`. So, like you said, with no need of data type validation.
But I can't see were/how to apply the same logic in `flatten` function. For now I should expect `dict` or `DataFrame`. But there is no way to `cast` (?) from one to another in a straight way as ``DataFrame` to be converted to a `dict` I must use the method `to_dict` which doesn exist for `dict`....
Other possibility is to convert what comes to `DataFrame`, but in this case, the tests for `dict` will fail.

I am going to far?

Reply:
My provocation had two functions with exactly the same name (`add`). One deals with arguments that are integer numbers. The other deals with arguments that are strings. 

Now replace in that sentence:
* `add` by `flatten`
* _integer numbers_ by _dictionaries_
* _strings_ by _Pandas DataFrame_

Reply:
> My provocation had two functions with exactly the same name (add). 

Is that possible? To have two functions with the same name?
... goin to check. I will be right back

Reply:
> Is that possible?

[Copy, paste and run it](https://github.com/FelipeSBarros/crossfire/issues/96#issuecomment-1960282545). Then you tell me : )

Reply:
Sorry, @cuducos I think I dind't get your point...

Reply:
https://github.com/FelipeSBarros/crossfire/assets/4732915/ce4d49b5-2bf1-4f59-bba0-e6d5fe1def3c

Reply:
Sorry I did what you show in the video, but had to left without time to write about.
After your comment I thought that it as posible to have two function with the same name but different evaluation.
So I try it ad, as I could see/understand the second function overwrite the first one.

Also I didn't get your point with the provocation. Some how your provocation called my attention to the idea of  "casting" the `data` inpiut to be flatten. This is why I did this tranformation: If `flatten` function already works with `dict`, to make the test pass once receving a `DataFrame`, is just a matter of transforming it to `dict`.
Of course, I believe that this would be the soluction fornow...
But now I am completely lost. ...And I am sorry about it.

Reply:
> So I try it and, as I could see/understand, the second function overwrite the first one.
> So I discarted this suposition...

`git commit` and `git push` or didn't happen.

Reply:
> `git commit` and `git push` or didn't happen.

Ah, I just tested the `add` function to "see" what happens, just like you did in the video.
If I understand right I should duplicate `flatten` function, the first to recieve `dict` and the second, to recieve `DataFrame`.
Right?

Reply:
@cuducos after much thinking, I decided to try again the test with your provocation. This time I recorded, got the courage to use nvim and added a print to identify the add function. But now I am more confused than before. I can see that this first function ocupied a different memory space then the second. What I don't know is wheter both still existing, and is being used according to the input...
 

https://github.com/FelipeSBarros/crossfire/assets/7208674/5d3c662e-eb15-41ae-ad4e-957708f07757

Reply:
Another test I did: I changed the restun adding a f-string to confirm if both functions still existing and being used according to the data input.
I can see that indeed the second function overwriten the first one, ocupying a noto memmory space. So event when calling `add` function with `interger` the last function in used, converting it to `int` event not being necessary.

https://github.com/FelipeSBarros/crossfire/assets/7208674/7f948693-1a6b-4faf-8e7e-a4fe26ceef10

Reply:
Reading back yout [commen](https://github.com/FelipeSBarros/crossfire/pull/103#discussion_r1501166414)


> My provocation had two functions with exactly the same name (`add`). One deals with arguments that are integer numbers. The other deals with arguments that are strings.
> 
> Now replace in that sentence:
> 
> * `add` by `flatten`
> * _integer numbers_ by _dictionaries_
> * _strings_ by _Pandas DataFrame_

Makes me think exactly what I dad before, and have mentioned [here](https://github.com/FelipeSBarros/crossfire/pull/103#discussion_r1501293832)

> Also, I didn't get your point with the provocation. Some how your provocation called my attention to the idea of "casting" the `data` input to be flatten. This is why I did this tranformation: If `flatten` function already works with `dict`, to make the test pass once receving a `DataFrame`, is just a matter of transforming it to `dict`.
> Of course, I believe that this would be the solution for now, to make the test pass and then to start improving...

If converting between `dict` and `DataFrame` was not your point, what was it?

Reply:
Can you commit and push _or_ share your code so I don't have to copy it from a video if I wanna reproduce, comment and tweak it?

Reply:
Ow! I just read that this experimental feature is removed and only used for type-checking! Oh my goddess, sorry I made you spend so much time on this dead-end! Totally my fault!

Reply:
Don't worry. It is a relieve to realize that I am not crazy.
But, just to confirm: the *experimental feature* you mention is the possibility of having two functions with different parameters/processes?